### PR TITLE
[Enhancement] Remove redundant management on open db

### LINKED ISSUE #8334: Remove redundant management on open db
FE uses Berkely DB to store Journal and maintain high availability. The `BDBEnvironment` class is responsible for opening a database instance and returning the handler of the database,  but it never closes the database initiatively. Only when a snapshot is made, databases are purged to the snapshot and are ready to clean, will the handler close passively.

The initial goal of holding all the handlers is probably to avoid frequently open databases, but it requires further thinking.

If the current FE is a master, two threads will open DB for now. 
* JournalWriter will open DB to append journals in sequence.
* Checkpointer will open DB to replay journals in sequence from all the rolled DBs. It will only reopen the database that JournalWriter closed for once, but since Checkpointer has no requirement on efficiency, the cost of reopening hardly matters.

If the current FE is a follower, there is only one thread that will open DB for now.
* Replayer will open DB to replay journals in sequence.

Besides, during the procedure of master/follower state change, FE will load the latest image and open DB to replay journals in sequence. 

For the above reasons, I think the management of opened DBs is redundant. Furthermore, it will bring out some bad consequences.

* If FE hasn't made checkpoints for a long time, either because the interval is set too long or because initiatively gives up on short of memory, there will be a leak of opened databases as the process runs.
* We can always get a lot of errors in the follower such as "Database xx has been forcibly closed in order to apply a replicated remove operation". This is possible because we leave the handler open while it was removed by the master.

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #8334 

## Problem Summary(Required) ：

1. Remove redundant management on open db in BDBEnvironment.
2. Close database initiatively in BDBJournal & BDBJournalCursor
3. In every main loop, Replayer will update names of local databases by calling refresh(), then keep reading increasing log id until it doesn't exist.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 13 / 22 (59.09%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJEJournal.java|4|10|40.00%|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|3|3|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 94 / 124 (75.81%)




#### file detail

|   |path|covered line|new line|coverage|not covered line detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|13|38|34.21%|[1493, 1500, 1502, 1503, 1505, 1508, 1510, 1511, 1513, 1514, 1703, 1704, 1706, 1707, 1747, 1749, 1750, 1751, 1752, 1755, 1756, 1757, 1758, 1764, 1765]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|5|8|62.50%|[473, 481, 482]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|66|68|97.06%|[93, 225]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJEJournal.java|10|10|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java [hunk 1]
```diff
@@ -41,50 +41,123 @@
...
-        this.dbNames = env.getDatabaseNames();
-        if (dbNames == null) {
-            throw new JournalException("failed to get db names!");
+    /**
+     * init journal cursor
+     * if toKey = -1(CUROSR_END_KEY), it will automatically search the end.
+     */
+    public static BDBJournalCursor getJournalCursor(BDBEnvironment env, long fromKey, long toKey)
+            throws JournalException, JournalInconsistentException, InterruptedException {
+        if (fromKey < 0 || toKey > 0 && toKey < fromKey || toKey <= 0 && toKey != JournalCursor.CUROSR_END_KEY) {
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java:92]: Could you add some parentheses? Too many conditions

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java [hunk 2]
```diff
@@ -465,100 +457,29 @@ public ReplicatedEnvironment getReplicatedEnvironment() {
...
-                openedDatabases.add(db);
-                LOG.info("successfully open new db {}", db);
-            } catch (Exception e) {
-                LOG.warn("catch an exception when open database {}", dbName, e);
-            }
-        } finally {
-            lock.writeLock().unlock();
-        }
-        return db;
+        return new CloseSafeDatabase(replicatedEnvironment.openDatabase(null, dbName, dbConfig));
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java:465]: As I analyzed in the description of this issue, there's no chance that a database is used multiple times or by multiple threads.

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java (hunk 1 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java]: Done.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 94 / 124 (75.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|13|38|34.21%|[1518, 1526, 1528, 1529, 1531, 1534, 1536, 1537, 1539, 1540, 1729, 1730, 1732, 1733, 1773, 1775, 1776, 1777, 1778, 1781, 1782, 1783, 1784, 1790, 1791]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|5|8|62.50%|[474, 482, 483]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|66|68|97.06%|[96, 228]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJEJournal.java|10|10|100.00%|[]|

**Review by gengjun-git — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 94 / 124 (75.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|13|38|34.21%|[1516, 1524, 1526, 1527, 1529, 1532, 1534, 1535, 1537, 1538, 1727, 1728, 1730, 1731, 1771, 1773, 1774, 1775, 1776, 1779, 1780, 1781, 1782, 1788, 1789]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|5|8|62.50%|[474, 482, 483]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|66|68|97.06%|[96, 228]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJEJournal.java|10|10|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 3]
```diff
@@ -1693,60 +1713,82 @@ protected synchronized void runOneCycle() {
...
-            return false;
-        }
-
-        LOG.info("replayed journal id is {}, replay to journal id is {}", replayedJournalId, newToJournalId);
+    /**
+     * Replay journal from replayedJournalId + 1 to toJournalId
+     * used by checkpointer/replay after state change
+     */
+    public boolean replayJournal(long toJournalId) throws JournalException {
+        LOG.info("replayed journal id is {}, replay to journal id is {}", replayedJournalId, toJournalId);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1721]: LOG.info("start to replay journal from {} to {}", replayedJournalId + 1, toJournalId)

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java [hunk 4]
```diff
@@ -465,100 +457,30 @@ public ReplicatedEnvironment getReplicatedEnvironment() {
...
-                LOG.warn("catch an exception when remove db:{}, this db does not exist", dbName, e);
-            }
-        } finally {
-            lock.writeLock().unlock();
+            LOG.info("begin to remove database {} from replicatedEnviroment", dbName);
+            // the first parameter null means auto-commit
+            replicatedEnvironment.removeDatabase(null, dbName);
+            LOG.info("remove database {} from replicatedEnviroment successfully", dbName);
+        } catch (DatabaseNotFoundException e) {
+            LOG.warn("catch an exception when remove db:{}, this db does not exist", dbName, e);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java:483]: LOG.warn("Exception: {} does not exist", dbName, e);

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java [hunk 5]
```diff
@@ -465,100 +457,30 @@ public ReplicatedEnvironment getReplicatedEnvironment() {
...
-            try {
-                LOG.info("begin to remove database {} from replicatedEnviroment", dbName);
-                // the first parameter null means auto-commit
-                replicatedEnvironment.removeDatabase(null, dbName);
-            } catch (DatabaseNotFoundException e) {
-                LOG.warn("catch an exception when remove db:{}, this db does not exist", dbName, e);
-            }
-        } finally {
-            lock.writeLock().unlock();
+            LOG.info("begin to remove database {} from replicatedEnviroment", dbName);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java:478]: What the purpose of this log?

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java [hunk 6]
```diff
@@ -41,50 +41,126 @@
...
+    }
+
+    protected BDBJournalCursor(BDBEnvironment env, long fromKey, long toKey) {
+        this.environment = env;
+        this.currentKey = fromKey;
+        this.toKey = toKey;
+    }
+
+    @Override
+    public void refresh() throws JournalException, JournalInconsistentException, InterruptedException {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java:131]: How about the refresh() performance()?

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java (hunk 3 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java]: Done

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java (hunk 4 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java]: Done

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java (hunk 5 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java]: removed

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java (hunk 6 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java:138]: I've added to the comment of the UT.

```
        // 2022-07-12 17:20:49 .. - call environment.getDatabaseNames() 1000000 times cost 6797 ms
        // 2022-07-12 18:56:41 .. - call environment.getDatabaseNames() 1000000 times cost 6892 ms
        // seems like an in-memory operation, only cost 6µs
```

**Comment by mofeiatwork** [comment]
run all

**Comment by padmejin** [comment]
run starrocks_admit_test

**Review by chaoyli — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
run starrocks_admit_test

**Pull request merged**